UI: Add status overlay for macOS dock icon

gxalpha opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
macOS equivalent of #3153.
Position of the icon on the bottom left matches the status bar.

Recording/Streaming:
<img width="52" alt="image" src="https://user-images.githubusercontent.com/59806498/156207459-e7eadd80-44b7-481f-a026-4ba5ec88329c.png">

Paused:
<img width="51" alt="image" src="https://user-images.githubusercontent.com/59806498/156207511-d2d4ba30-246e-4efb-be0a-cb1a32c7258d.png">

Normal (existing):
<img width="52" alt="image" src="https://user-images.githubusercontent.com/59806498/156207545-f4b23792-4ab6-49a0-aac6-5b46185f681b.png">


Doesn't rely on QtMacExtras, which means it will survive the transition to Qt6 (which the existing Windows implementation won't, that needs reworking before Qt6).
Implementation for other Unix-like is not included, the functions are just added so that everything compiles.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
This is a nice QOL addition on Windows and should be available on more platforms.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
Compiled on macOS 12.3, Windows CI and Linux CI.
Made sure Windows implementation still works (Windows 11).
Tested that macOS works as intended, see the screenshot above.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
- Tweak (non-breaking change to improve existing functionality)
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

tytan652 commented:
I don't think it's implementable on Linux on all Desktop Environments, most of them just use the SVG in the `.desktop` file.

This shall maybe be macOS and Windows only in the code.

gxalpha commented:
> I don't think it's implementable on Linux on all Desktop Environments

Does this mean it's possible in *some* environments?
> This shall maybe be macOS and Windows only in the code.

It essentially is. I really hate the idea of ifdefing all the statements again, makes the code hard to read. I personally think having these empty functions in platform-x11 is fine; and they'll likely get optimized away anyways.

tytan652 commented:
> Does this mean it's possible in _some_ environments?

I seriously don't know, I just consider that not all DEs are not made the same way.

But I think we should avoid per DEs features on Linux and try to provide the same amount feature on all DEs.

tytan652 commented on UI/platform-x11.cpp:
```diff
@@ -257,3 +257,8 @@ bool SetDisplayAffinitySupported(void)
 	return false;
 }
+
+// Not implemented yet
```
It's this comment that made me write my first comment.

gxalpha commented:
I disagree. Why should one DE hold others back? We're generally happy to introduce features to specific operating systems only (a prime example is this exact feature), I don't see why this should be different for specific DE's. That is to say, if someone were to come and to implement it for only one DE, straight up saying "all or none" seems wrong to me.

But that's a discussion that's not in scope of this PR.
If it's just about the comment, I can remove that if it would be a blocker for you.

tytan652 commented:
> If it's just about the comment, I can remove that if it would be a blocker for you.

You can leave it, just wanted to tell my opinion that we should try to treat DEs equally. And there is also the Flatpak that ignore in which DE it is run.

For Windows there is only one "DE", macOS same. But on Linux, many of them.

You are right this discussion is out of the scope of this PR, sorry for the inconvenience.

gxalpha commented:
No problem, thanks for bringing it up!
I guess it can get discussed again should someone be doing a Linux implementation.

PatTheMav commented on UI/platform-osx.mm:
```diff
@@ -255,6 +255,51 @@ bool ProcessIsRosettaTranslated()
+}
+- (void)drawRect:(NSRect)dirtyRect
+{
+	CGSize size = dirtyRect.size;
```
@gxalpha How often is this called? If the icon (including the status icon) only changes on actual changes of the app's status, then redrawing the overlay on every draw call of the composer seems wasteful to me.

gxalpha replied:
It has to be called manually with `[tile display]`, so it's only called when `TaskbarOverlaySetIcon` is.

PatTheMav commented on UI/platform-windows.cpp:
```diff
@@ -469,3 +474,26 @@ bool IsRunningOnWine()
+
+#if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
+QWinTaskbarButton *taskBtn;
+void TaskbarOverlayInit()
```
If this PR also adds functionality to Windows, it should either be renamed or the changes decoupled into another PR.

Right now Jim (and I) were under the assumption this only adds functionality to macOS, but if it also changes other platforms I cannot fully review+merge this on my own.

gxalpha replied:
This functionality already exists for Windows. It was only moved from window-basic-main to the platform file, functionally was not changed.

PatTheMav replied:
Ah right, you just moved the code around. Did you check the functionality on Windows?

gxalpha replied:
I downloaded CI builds on my fork and tested it still works for recordings

PatTheMav commented:
Code looks fine to me, but I'm not entirely happy with the composition of the icon and the overlays - the record icon is too similar to the notification badge and detached elements of icons usually are put in the bottom right corner (compare: Photo Booth, Preview, Disk Utility, Chess).

Maybe if both icons are contained in another shape and this shape is the "common" element between both it allows it to be distinct from the notification badge and also make the <pause> icon feel less "floaty".

gxalpha commented:
> detached elements of icons usually are put in the bottom right corner

I deliberately put it in the bottom left corner since with the Windows implementation, users semi-regularly complained that it doesn't match the taskbar icon. If you want me to, I can put it in the bottom right, sure.

> Maybe if both icons are contained in another shape and this shape is the "common" element between both it allows it to be distinct from the notification badge and also make the icon feel less "floaty".

I'd need a mockup to understand what you mean, I'm pretty bad at visualizing things.

PatTheMav commented:
I was thinking along the lines of this - just quickly whipped-up examples:

Recording Active

![example_obs_icon_record](https://user-images.githubusercontent.com/9436503/158234427-af8ee25b-8220-44cc-84ce-cfa5a29a365c.png)

Recording Paused

![example_obs_icon_pause](https://user-images.githubusercontent.com/9436503/158234451-fa8ff2cb-4fcf-46c1-91d1-d53c0705282a.png)

Colors/stroke widths obviously needs tweaking, but I guess you get the idea.

WizardCM commented:
Would you be willing to split off the changes that standardise/version-gate the Windows function into a separate commit?

gxalpha commented:
Split off into #6371. Will be rebase this one once 6371 is merged.

PatTheMav commented:
Has the champ already spoken on this?

gxalpha commented:
no

MisutaaAsriel commented:
Does this take into account PR #6009? macOS sets the dock icon by default via the PLIST, not by code, and users (such as myself) may wish to change the user icon in the dock via Finder (which changes to the icon by code override).

_Personally_, the option to enable status in the dock via settings would be preferable to being forced to use this. **I genuinely do not like dock icons changing themselves, or attempting to behave like a status bar**.

gxalpha commented:
> Does this take into account PR https://github.com/obsproject/obs-studio/pull/6009?

While I didn’t test it, I wrote it in a way that should respect any custom icons.

gxalpha commented:
Rebased to fix conflicts introduced in #6371.
Confirmed that custom icons are respected and do not get overridden, with the overlay correctly placed on top.

gxalpha commented:
As a note, this still requires a comment and/or design from @Warchamp7 regarding https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1062185211 / https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1067134057

Warchamp7 commented:
> As a note, this still requires a comment and/or design from @Warchamp7 regarding [#6060 (comment)](https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1062185211) / [#6060 (comment)](https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1067134057)

Are there any images of how this looks currently? I'd personally like to match Windows having this indicator bottom left, but I defer to @PatTheMav as resident Mac person if he thinks having it bottom right would be more important.

Visually if we want a shared shape for the icons to sit on, I'd propose a white rounded square matching the radius of the whole icon, and removing the colored outline (or at least the drop shadow on it)

PatTheMav commented:
> > As a note, this still requires a comment and/or design from @Warchamp7 regarding [#6060 (comment)](https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1062185211) / [#6060 (comment)](https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1067134057)
> 
> Are there any images of how this looks currently? I'd personally like to match Windows having this indicator bottom left, but I defer to @PatTheMav as resident Mac person if he thinks having it bottom right would be expected on Mac and more important than having parity with Windows status location.
> 
> Visually if we want a shared shape for the icons to sit on, I'd propose a white rounded square matching the radius of the whole icon, and removing the colored outline (or at least the drop shadow on it)
> 
> Edit: Based off Pat's mockup
> 
> ![Mac Example](https://user-images.githubusercontent.com/1554753/175789480-2eac5b63-f43a-4d70-8210-36e2b0703836.jpg)
> 
> [Mac Record Top](https://user-images.githubusercontent.com/1554753/175789488-a70707e8-6356-47b3-bcd2-2daf38333a72.png) [Mac Pause Top](https://user-images.githubusercontent.com/1554753/175789489-f908b11e-85ec-454c-96f9-089072caf9b3.png)

In general dock icon overlays are uncommon in macOS (because that's what menu bar icons and the control center are for), which also means there aren't any rules for this.

If there's bigger support for having the icon in the bottom left, I'm fine with putting it there. The square variant I'm also fine with. Personally I'd like to be able to disable this, but I guess @gxalpha has to adapt the code to use macOS-specific overlay icons first anyway.

gxalpha commented:
Updated the icons to macOS specific ones and moved them to the bottom right.
Also made them a tiny bit bigger because with the ones posted above, they are very small and not really visible.

<img width="123" alt="image" src="https://user-images.githubusercontent.com/59806498/179866802-e4d7d117-e7fb-4f55-a909-f0075d530b76.png">
<img width="121" alt="image" src="https://user-images.githubusercontent.com/59806498/179866817-ee991b6c-0825-43fa-8e08-ce2bcf3c16d2.png">

MisutaaAsriel commented:
> In general dock icon overlays are uncommon in macOS (because that's what menu bar icons and the control center are for), which also means there aren't any rules for this.
> 
> If there's bigger support for having the icon in the bottom left, I'm fine with putting it there. The square variant I'm also fine with. Personally I'd like to be able to disable this, but I guess @gxalpha has to adapt the code to use macOS-specific overlay icons first anyway.

Wouldn't the doc overlay be more appropriate as a circle in the top right? This is where macOS currently puts any form of "status" icon, **e.g.** notification icons. This would at least match the look and feel of native system behavior.

Mockup based on system appearance:
![obs_live](https://user-images.githubusercontent.com/8140068/180063162-99362b86-5377-4dd3-8bcd-1a5554d395db.png) ![obs_paused](https://user-images.githubusercontent.com/8140068/180063200-076ab03e-9c8c-4d49-bae5-8036ce1cd442.png)

gxalpha commented:
If you look at https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1062185211, it was specifically made to *not* look like a notification. That being said, I'm personally open to any position, as long as the rest of the team agrees on it.

MisutaaAsriel commented:
> If you look at [#6060 (comment)](https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1062185211), it was specifically made to _not_ look like a notification. That being said, I'm personally open to any position, as long as the rest of the team agrees on it.

I would actually argue that using a badge similar to the notification badge would be ideal. Some applications throw "generic" notifications that either show just a red circle, or a red circle with a dot, as a way of grabbing user attention.

When OBS is recording, if there's going to be a dock notification, it is, well, **notifying** you that OBS is recording.

But to be fair, **Pat has a point** — [Apple says it's bad practice to mimic badges in apps, because badges can be turned off by the system](https://developer.apple.com/design/human-interface-guidelines/components/system-experiences/notifications). — If a user tried disabling it in Notifications, only for it to still show, they'd probably be confused. So there is an argument for making them *dissimilar* to badges.

**Being said**, the visibility of the current versions *leaves much to be desired*, and can be argued as an accessibility issue. In [#6060 (comment 1189657207)](https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1189657207), it's really hard to make out the status of the pause icon, and in general the visibility to me is really poor. This is why the solid color background with white (for dark colors) or black (for bright colors) glyph seems to make more sense.

As it's not standard practice anyways *to* display statuses on the dock, we're kinda throwing the rulebook out anyways. Likewise, the option to enable, disable these in OBS settings **should** be present, idealistically, in a final implementation, ergo I would argue using badging is not inappropriate in this particular case. The whole point of the icons are to grab user attention, so visibility and accessibility should take priority, imho.

You also have to recognize that a record icon looks exactly like the generic badge icon, so any changes to it to make it distinct are also going to interfere with visibility.

It really could go either way. I'm not fond of how it looks now in these screenshots, but I admit that while badging does increase visibility and recognizability, *it's a minor faux pas against the Human Interface Guidelines.*

...just, not any more than status icons in the dock in general are.

PatTheMav commented:
FWIW I've been arguing that it is standard practise in macOS to put such status displays into the menu bar icon and _not_ on the dock icon, which avoids any such discussions (especially as that icon has all the desired functionality already).

If it has to be on the dock icon (and I agree that best-case it should be a setting that users can disable), the top right corner is not available because it's already taken by the OS.

MisutaaAsriel commented:
> If it has to be on the dock icon (and I agree that best-case it should be a setting that users can disable), the top right corner is not available because it's already taken by the OS.

Well, if the OS prevents top-right display, I really cannot argue with that.

*Being said*, as it stands, I still find the icons present an accessibility issue. Ideally, if there are going to be icons, they should meet criteria such as:

- Easy to discern
- High contrast
- Low detail
- Large glyphs

The icons as they are are rather small in nature, and in the pause icon in particular, places a light colored glyph atop a white background. This harms legibility and defeats the purpose of a dock icon, *as is*. I'm not opposed to them, but the current design is rather counterintuitive.

Taking in the feedback previous of this conversation however, I've attached some sample icons which are equal in canvas resolution to the icons in this commit\*, but seek to increase visibility, whilst taking design cues from the previous rendition, and seeking to blend them in with the default OS aesthetic without being too similar to notifications. Also: squircles.

![image](https://user-images.githubusercontent.com/8140068/180104192-6a3916d5-9b3f-4e85-88fa-09aa15749e7a.png) ![image](https://user-images.githubusercontent.com/8140068/180104206-a27ef456-7270-4159-b237-10979e0a49bd.png)

Also attached are icon samples of what it might look like; I'm assuming whatever method is being used would support the full icon "canvas" or image size, and not just to the border of the icon itself.

![image](https://user-images.githubusercontent.com/8140068/180104347-3f7a28d0-9c72-4844-b070-38eeda84bcbc.png) ![image](https://user-images.githubusercontent.com/8140068/180104418-d84d792d-afa1-4be3-9d62-d83aef4a6261.png)

\**The actual icon is about 96px, similar to other dock icon overlays **like** notifications, with a slight drop-shadow, resulting in about a 10 pixel difference (112 vs 122). This is to be consistent with the OS, not to copy it.*

(Sorry if I come off as annoying btw; I'm passionate about both software design and streaming, and as an OBS user and Mac user, I want to help wherever I am able.)

PatTheMav commented:
> > If it has to be on the dock icon (and I agree that best-case it should be a setting that users can disable), the top right corner is not available because it's already taken by the OS.
> 
> Well, if the OS prevents top-right display, I really cannot argue with that.
> 

The OS doesn't prevent that, as we're in essence replacing the dock item with an image that has the badge "baked-in". If there'd be a notification, macOS would in essence draw over our status icon, as it doesn't care for icon decorations by apps.

It's for that reason that I deem that corner "unavailable".

As for the visual design I defer to Warchamp's judgement.

Warchamp7 commented:
> > If it has to be on the dock icon (and I agree that best-case it should be a setting that users can disable), the top right corner is not available because it's already taken by the OS.
> 
> Well, if the OS prevents top-right display, I really cannot argue with that.
> 
> _Being said_, as it stands, I still find the icons present an accessibility issue. Ideally, if there are going to be icons, they should meet criteria such as:
> 
> * Easy to discern
> * High contrast
> * Low detail
> * Large glyphs
> 
> The icons as they are are rather small in nature, and in the pause icon in particular, places a light colored glyph atop a white background. This harms legibility and defeats the purpose of a dock icon, _as is_. I'm not opposed to them, but the current design is rather counterintuitive.
> 
> Taking in the feedback previous of this conversation however, I've attached some sample icons (mockups, more or less) which are equal in canvas resolution to the icons in this commit*, but seek to increase visibility, whilst taking design cues from the previous rendition, and seeking to blend them in with the default OS aesthetic without being too similar to notifications. Also: squircles.
> 
> ![image](https://user-images.githubusercontent.com/8140068/180105601-ce0e9a4e-96e4-428a-bd05-2a978b0df9bb.png) ![image](https://user-images.githubusercontent.com/8140068/180105619-f9845a58-8ecb-4f46-bed4-993967a5f531.png) (_Updated for graphical error_)
> 
> Also attached are icon samples of what it might look like; I'm assuming whatever method is being used would support the full icon "canvas" or image size, and not just to the border of the icon itself.
> 
> ![image](https://user-images.githubusercontent.com/8140068/180104347-3f7a28d0-9c72-4844-b070-38eeda84bcbc.png) ![image](https://user-images.githubusercontent.com/8140068/180104418-d84d792d-afa1-4be3-9d62-d83aef4a6261.png)
> 
> *_The actual icon is about 96px, similar to other dock icon overlays **like** notifications, with a slight drop-shadow, resulting in about a 10 pixel difference (112 vs 122). This is to be consistent with the OS, not to copy it._
> 
> (Sorry if I come off as annoying btw; I'm passionate about both software design and streaming, and as an OBS user and Mac user, I want to help wherever I am able.)

I'm good with these icons but can you adjust the colors to #ff001e for Recording and #fecb34 for Pause?

MisutaaAsriel commented:
> @MisutaaAsriel I'm good with these icons but can you adjust the colors to #ff001e for Recording and #fecb34 for Pause?

Those icons were a little rough around the edges; more concept than productionr ready. But if you're serious about using them, here's some somewhat cleaned up versions with the changes requested:

![image](https://user-images.githubusercontent.com/8140068/180675099-79823206-7cfe-46e9-a4f0-492c396ff2ba.png) ![image](https://user-images.githubusercontent.com/8140068/180674694-f1021789-2526-4e82-bc8f-a641eaf22e21.png)

gxalpha commented:
Newest update (with the icons from https://github.com/obsproject/obs-studio/pull/6060#issuecomment-1193447442):
<img width="59" alt="image" src="https://user-images.githubusercontent.com/59806498/180785741-8a432422-421c-4be2-b254-3bbb6c819bcf.png">
<img width="57" alt="image" src="https://user-images.githubusercontent.com/59806498/180785778-eadd4fef-5d76-493d-bd93-a23edc94f296.png">

mufunyo commented:
> ![image](https://user-images.githubusercontent.com/8140068/180675099-79823206-7cfe-46e9-a4f0-492c396ff2ba.png) ![image](https://user-images.githubusercontent.com/8140068/180674694-f1021789-2526-4e82-bc8f-a641eaf22e21.png)

Compared to Warchamp's version, the outer border is now mismatched with the inner border. This looks extremely off.

MisutaaAsriel commented:
> Compared to Warchamp's version, the outer border is now mismatched with the inner border. This looks extremely off.

I'll try making changes to it; now that you mention it you are correct, though i didn't see it at the time. Thank you for pointing that out.

MisutaaAsriel commented:
@Warchamp7 @mufunyo @gxalpha Reworked them in a proper vector graphics tool; This should close up any visual discrepancies caused by my other program and be more consistent. This also allows me to provide SVGs.

![Icon](https://user-images.githubusercontent.com/8140068/181131914-3f088203-6283-4050-ba74-78eb79a4557d.png) ![Icon](https://user-images.githubusercontent.com/8140068/181131539-4fbd7ad5-b513-4370-a615-8504c7a8756f.png)

SVG:
![Icon](https://user-images.githubusercontent.com/8140068/181132256-1541591e-6321-4c6d-9f99-759f8dd5a984.svg) ![Icon2](https://user-images.githubusercontent.com/8140068/181132487-3622b2a4-f1cd-42dd-a7b8-3165b7faf570.svg)

gxalpha commented:
In case you're wondering, we merged it as-is because it didn't care too much that the icons were imperfect since they are very small when shown anyways, and we didn't want to hold this PR up any more.
I can just update them to this new version.

MisutaaAsriel commented:
> In case you're wondering, we merged it as-is because it didn't care too much that the icons were imperfect since they are very small when shown anyways, and we didn't want to hold this PR up any more.
> I can just update them to this new version, thanks for making that.

Noted. And thank you for being gracious enough to update them. No matter how small, I'd be embarrassed if something I did was visually flawed, and this _is_ a Mac feature; Mac users might be more inclined to notice such things, especially at larger dock sizes.

gxalpha commented:
Pushed the new icons. Again, thanks for making them.
